velox/functions/sparksql/DateTimeFunctions.h

```diff
@@ -408,4 +408,47 @@ struct DayOfYearFunction {
     result = getDayOfYear(getDateTime(date));
   }
 };
+
+template <typename T>
+struct DateFormatFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  const date::time_zone* sessionTimeZone_ = nullptr;
+  std::shared_ptr<DateTimeFormatter> formatter_;
+  bool isConstFormat_ = false;
+
+  FOLLY_ALWAYS_INLINE void setFormatter(const arg_type<Varchar>* formatString) {
+    if (formatString != nullptr) {
+      formatter_ = buildJodaDateTimeFormatter(
+          std::string_view(formatString->data(), formatString->size()));
+      isConstFormat_ = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<Timestamp>* /*timestamp*/,
+      const arg_type<Varchar>* formatString) {
+    sessionTimeZone_ = getTimeZoneFromConfig(config);
+    setFormatter(formatString);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Timestamp>& timestamp,
+      const arg_type<Varchar>& formatString) {
+    if (!isConstFormat_) {
+      formatter_ = buildJodaDateTimeFormatter(
+          std::string_view(formatString.data(), formatString.size()));
+    }
+
+    auto formattedResult = formatter_->format(timestamp, sessionTimeZone_);
+    auto resultSize = formattedResult.size();
+    result.resize(resultSize);
+    if (resultSize != 0) {
+      std::memcpy(result.data(), formattedResult.data(), resultSize);
```

Comment:
Is there a way to avoid this copy? Can we compute the size of the result first, then 'format' directly into the result?

Reply:
Yes, we can avoid this copy. We may have to change formatter's `format` method to support passing a result argument by reference. Such change will impact on other code where it is called. Should I do this optimization in this PR or in a separate PR?

Reply:
Maybe create a separate PR to change formatter's format method, then updating this one once it lands.

Reply:
Sounds good. I will create a separate to do that. Thanks!

Reply:
> Is there a way to avoid this copy? Can we compute the size of the result first, then 'format' directly into the result?

Hi @mbasmanova, I tried to precompute result size according to the given date format arg. But I found the result size also depends on each timestamp input. For example, the result for `dateFormat("19690-12-31", "yyyy-MM-dd")` is `19690-12-31`, whose size is larger than "yyyy-MM-dd". Do you have any suggestion? Thanks!

Reply:
@PHILO-HE I see a couple of options: (1) pass timestamp value to a function that computes the result; (2) allow format function to resize the result.

Reply:
@PHILO-HE @mbasmanova Could we instead make the date formatter object take a complex type writer (sink/ostream or similar abstraction), since these things are in the end of the day already backed by a larger FlatVector<StringView>?